<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Topic 2-02: Implement ORA from stratch • GSEAtraining</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Topic 2-02: Implement ORA from stratch">
<meta property="og:description" content="GSEAtraining">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GSEAtraining</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/topic1_00_format.html">Topic 1-00: Representation of gene sets in R</a>
    </li>
    <li>
      <a href="../articles/topic1_01_GO.html">Topic 1-01: Get GO gene sets</a>
    </li>
    <li>
      <a href="../articles/topic1_02_kegg.html">Topic 1-02: Get pathways from KEGG</a>
    </li>
    <li>
      <a href="../articles/topic1_03_msigdbr.html">Topic 1-03: Get pathways from MSigDB</a>
    </li>
    <li>
      <a href="../articles/topic1_04_Reactome.html">Topic 1-04: Get pathways from Reactome</a>
    </li>
    <li>
      <a href="../articles/topic1_05_BioCarta.html">Topic 1-05: Get pathways from BioCarta</a>
    </li>
    <li>
      <a href="../articles/topic1_06_UniProtKeywords.html">Topic 1-06: Get gene sets from UniProt Keywords</a>
    </li>
    <li>
      <a href="../articles/topic1_07_more_gene_sets.html">Topic 1-07: Get GO/KEGG gene sets for other organisms</a>
    </li>
    <li>
      <a href="../articles/topic1_08_ortholog.html">Topic 1-08: Generate gene sets for other organisms by mapping to orthologues</a>
    </li>
    <li>
      <a href="../articles/topic1_09_gene_id.html">Topic 1-09: Gene ID conversion</a>
    </li>
    <li>
      <a href="../articles/topic2_01_ora_single_gene_set.html">Topic 2-01: ORA on a single gene set</a>
    </li>
    <li>
      <a href="../articles/topic2_02_implement_ora.html">Topic 2-02: Implement ORA from stratch</a>
    </li>
    <li>
      <a href="../articles/topic2_03_clusterProfiler_ora.html">Topic 2-03: Use clusterProfiler for ORA</a>
    </li>
    <li>
      <a href="../articles/topic2_04_implement_gsea.html">Topic 2-04: Implement GSEA from scratch</a>
    </li>
    <li>
      <a href="../articles/topic2_05_clusterProfiler_gsea.html">Topic 2-05: Use clusterProfiler for GSEA</a>
    </li>
    <li>
      <a href="../articles/topic2_06_compare_ORA_GSEA.html">Topic 2-06: Compare ORA and GSEA</a>
    </li>
    <li>
      <a href="../articles/topic3_01_online_GREAT.html">Topic 3-01: Online GREAT analysis</a>
    </li>
    <li>
      <a href="../articles/topic3_02_local_GREAT.html">Topic 3-02: Local GREAT analysis</a>
    </li>
    <li>
      <a href="../articles/topic3_03_goseq.html">Topic 3-03: GOseq</a>
    </li>
    <li>
      <a href="../articles/topic4_01_CePa.html">Topic 4-01: CePa</a>
    </li>
    <li>
      <a href="../articles/topic4_02_GSVA.html">Topic 4-02: GSVA</a>
    </li>
    <li>
      <a href="../articles/topic5_01_similarity.html">Topic 5-01: Similarities of terms</a>
    </li>
    <li>
      <a href="../articles/topic5_02_simplifyEnrichment.html">Topic 5-02: simplifyEnrichment</a>
    </li>
    <li>
      <a href="../articles/topic6_01_visualization.html">Topic 6-01: Visualization</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/GSEAtraining/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Topic 2-02: Implement ORA from stratch</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu <a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a>
</h4>
            
            <h4 data-toc-skip class="date">2024-01-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/GSEAtraining/blob/HEAD/vignettes/topic2_02_implement_ora.Rmd" class="external-link"><code>vignettes/topic2_02_implement_ora.Rmd</code></a></small>
      <div class="hidden name"><code>topic2_02_implement_ora.Rmd</code></div>

    </div>

    
    
<p>Implementing ORA is very simple. As long as genes and gene sets are already in the correct ID types and formats, the implementation of ORA only needs several lines.</p>
<p>Three inputs for ORA:</p>
<ul>
<li>a vector of (DE) genes: <code>genes</code>
</li>
<li>a list of vectors where each vector contains genes in a gene set: <code>gene_sets</code>
</li>
<li>a vector of background genes: <code>universe</code>
</li>
</ul>
<p>We calculate the following numbers:</p>
<p>Number of total genes:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_universe</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">universe</span><span class="op">)</span></span></code></pre></div>
<p>Number of DE genes:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<p>Sizes of gene sets:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gene_sets</span>, <span class="va">length</span><span class="op">)</span></span></code></pre></div>
<p>Number of DE genes in gene sets:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gene_sets</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We rename <code>n_genes</code> to <code>k</code>:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">k</span> <span class="op">=</span> <span class="va">n_genes</span></span></code></pre></div>
<style>
.table {
width:400px;
}
</style>
<p>We put these variables into the 2x2 contigency table:</p>
<table class="table">
<thead><tr class="header">
<th></th>
<th align="right">in the set</th>
<th align="right">not in the set</th>
<th align="right">total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>DE</td>
<td align="right"><code>x</code></td>
<td align="right">-</td>
<td align="right"><code>k</code></td>
</tr>
<tr class="even">
<td>not DE</td>
<td align="right">-</td>
<td align="right">-</td>
<td align="right">-</td>
</tr>
<tr class="odd">
<td>total</td>
<td align="right"><code>m</code></td>
<td align="right">-</td>
<td align="right"><code>n_universe</code></td>
</tr>
</tbody>
</table>
<p>Then we can calculate p-values from the hypergeometric distribution as:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/Hypergeometric.html" class="external-link">phyper</a></span><span class="op">(</span><span class="va">x</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">m</span>, <span class="va">n_universe</span> <span class="op">-</span> <span class="va">m</span>, <span class="va">k</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>which is the same as:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/Hypergeometric.html" class="external-link">phyper</a></span><span class="op">(</span><span class="va">x</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">k</span>, <span class="va">n_universe</span> <span class="op">-</span> <span class="va">k</span>, <span class="va">m</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>Let’s put all these code into a function:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ora</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">genes</span>, <span class="va">gene_sets</span>, <span class="va">universe</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">n_universe</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">universe</span><span class="op">)</span></span>
<span>    <span class="va">n_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gene_sets</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gene_sets</span>, <span class="va">length</span><span class="op">)</span></span>
<span>    <span class="va">n</span> <span class="op">=</span> <span class="va">n_universe</span> <span class="op">-</span> <span class="va">m</span></span>
<span>    <span class="va">k</span> <span class="op">=</span> <span class="va">n_genes</span></span>
<span>    </span>
<span>    <span class="va">p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Hypergeometric.html" class="external-link">phyper</a></span><span class="op">(</span><span class="va">x</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">m</span>, <span class="va">n</span>, <span class="va">k</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>        gene_set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span>,</span>
<span>        hits <span class="op">=</span> <span class="va">x</span>,</span>
<span>        n_genes <span class="op">=</span> <span class="va">k</span>,</span>
<span>        n_gs <span class="op">=</span> <span class="va">m</span>,</span>
<span>        n_total <span class="op">=</span> <span class="va">n_universe</span>,</span>
<span>        p_value <span class="op">=</span> <span class="va">p</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>We can further improve the function by:</p>
<ol style="list-style-type: decimal">
<li>set the default of <code>universe</code> to the total genes in <code>gene_sets</code>
</li>
<li>intersect <code>genes</code> and <code>gene_sets</code> to <code>universe</code>. Note this step also removes duplicated genes</li>
<li>add adjusted log2 fold enrichment and p-values</li>
<li>order the result table by the significance</li>
</ol>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ora</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">genes</span>, <span class="va">gene_sets</span>, <span class="va">universe</span> <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">universe</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">universe</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="va">universe</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">universe</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="co"># restrict in universe</span></span>
<span>    <span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">genes</span>, <span class="va">universe</span><span class="op">)</span></span>
<span>    <span class="va">gene_sets</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">gene_sets</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">universe</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">n_universe</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">universe</span><span class="op">)</span></span>
<span>    <span class="va">n_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span>
<span>    </span>
<span>    <span class="va">x</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gene_sets</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">genes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">m</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gene_sets</span>, <span class="va">length</span><span class="op">)</span></span>
<span>    <span class="va">n</span> <span class="op">=</span> <span class="va">n_universe</span> <span class="op">-</span> <span class="va">m</span></span>
<span>    <span class="va">k</span> <span class="op">=</span> <span class="va">n_genes</span></span>
<span>    </span>
<span>    <span class="va">p</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Hypergeometric.html" class="external-link">phyper</a></span><span class="op">(</span><span class="va">x</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">m</span>, <span class="va">n</span>, <span class="va">k</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>        gene_set <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span>,</span>
<span>        hits <span class="op">=</span> <span class="va">x</span>,</span>
<span>        n_genes <span class="op">=</span> <span class="va">k</span>,</span>
<span>        n_gs <span class="op">=</span> <span class="va">m</span>,</span>
<span>        n_total <span class="op">=</span> <span class="va">n_universe</span>,</span>
<span>        log2fe <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">x</span><span class="op">*</span><span class="va">n_universe</span><span class="op">/</span><span class="va">k</span><span class="op">/</span><span class="va">m</span><span class="op">)</span>,</span>
<span>        p_value <span class="op">=</span> <span class="va">p</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="va">df</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">p_value</span>, <span class="st">"BH"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">=</span> <span class="va">df</span><span class="op">$</span><span class="va">gene_set</span></span>
<span>    <span class="va">df</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">p_adjust</span>, <span class="va">df</span><span class="op">$</span><span class="va">p_value</span><span class="op">)</span>, ,drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Let’s try our <code><a href="../reference/ora.html">ora()</a></code> with 1000 random genes on the MSigDB hallmark gene sets. <code><a href="../reference/msigdb.html">get_msigdb()</a></code> is from the <strong>GSEAtraining</strong> package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: AnnotationDbi</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Filter, Find, Map, Position, Reduce, anyDuplicated, aperm, append,</span></span>
<span><span class="co">##     as.data.frame, basename, cbind, colnames, dirname, do.call,</span></span>
<span><span class="co">##     duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,</span></span>
<span><span class="co">##     lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,</span></span>
<span><span class="co">##     pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,</span></span>
<span><span class="co">##     tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Biobase</span></span></code></pre>
<pre><code><span><span class="co">## Welcome to Bioconductor</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Vignettes contain introductory material; view with</span></span>
<span><span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## Warning: package 'S4Vectors' was built under R version 4.3.2</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:utils':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     findMatches</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     I, expand.grid, unname</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/GSEAtraining" class="external-link">GSEAtraining</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'GSEAtraining'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked _by_ '.GlobalEnv':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     ora</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="../reference/msigdb.html">get_msigdb</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"2023.2.Hs"</span>, collection <span class="op">=</span> <span class="st">"h.all"</span><span class="op">)</span></span>
<span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/random_genes.html">random_genes</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="fl">1000</span>, <span class="st">"ENTREZID"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="../reference/ora.html">ora</a></span><span class="op">(</span><span class="va">genes</span>, <span class="va">gs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                      gene_set hits n_genes n_gs</span></span>
<span><span class="co">## HALLMARK_HEDGEHOG_SIGNALING       HALLMARK_HEDGEHOG_SIGNALING    5     214   36</span></span>
<span><span class="co">## HALLMARK_XENOBIOTIC_METABOLISM HALLMARK_XENOBIOTIC_METABOLISM   16     214  200</span></span>
<span><span class="co">## HALLMARK_ANDROGEN_RESPONSE         HALLMARK_ANDROGEN_RESPONSE    9     214  101</span></span>
<span><span class="co">## HALLMARK_KRAS_SIGNALING_DN         HALLMARK_KRAS_SIGNALING_DN   15     214  200</span></span>
<span><span class="co">## HALLMARK_MYOGENESIS                       HALLMARK_MYOGENESIS   15     214  200</span></span>
<span><span class="co">## HALLMARK_APOPTOSIS                         HALLMARK_APOPTOSIS   12     214  161</span></span>
<span><span class="co">##                                n_total    log2fe    p_value  p_adjust</span></span>
<span><span class="co">## HALLMARK_HEDGEHOG_SIGNALING       4384 1.5085682 0.02899916 0.6219054</span></span>
<span><span class="co">## HALLMARK_XENOBIOTIC_METABOLISM    4384 0.7127089 0.03354442 0.6219054</span></span>
<span><span class="co">## HALLMARK_ANDROGEN_RESPONSE        4384 0.8682786 0.05667660 0.6219054</span></span>
<span><span class="co">## HALLMARK_KRAS_SIGNALING_DN        4384 0.6195995 0.06219054 0.6219054</span></span>
<span><span class="co">## HALLMARK_MYOGENESIS               4384 0.6195995 0.06219054 0.6219054</span></span>
<span><span class="co">## HALLMARK_APOPTOSIS                4384 0.6106107 0.09262817 0.6763795</span></span></code></pre>
<p>The only restriction of <code><a href="../reference/ora.html">ora()</a></code> is the ID type for <code>genes</code> should be the same as in <code>gene_sets</code>.</p>
<div class="section level2">
<h2 id="practice">Practice<a class="anchor" aria-label="anchor" href="#practice"></a>
</h2>
<div class="section level3 tabset">
<h3 class="tabset" id="section"></h3>
<div class="section level4">
<h4 id="practice-1">Practice 1<a class="anchor" aria-label="anchor" href="#practice-1"></a>
</h4>
<p>The file <code>de.rds</code> contains results from a differential expression (DE) analysis.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"de.rds"</span>, package <span class="op">=</span> <span class="st">"GSEAtraining"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">de</span> <span class="op">=</span> <span class="va">de</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"symbol"</span>, <span class="st">"p_value"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">de</span> <span class="op">=</span> <span class="va">de</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">de</span><span class="op">$</span><span class="va">p_value</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">de</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     symbol      p_value</span></span>
<span><span class="co">## 1   TSPAN6 0.0081373337</span></span>
<span><span class="co">## 2     TNMD 0.0085126916</span></span>
<span><span class="co">## 3     DPM1 0.9874656931</span></span>
<span><span class="co">## 4    SCYL3 0.2401551938</span></span>
<span><span class="co">## 5 C1orf112 0.8884603214</span></span>
<span><span class="co">## 6      FGR 0.0006134791</span></span></code></pre>
<p>Take the <code>symbol</code> and <code>p_value</code> columns, set cutoff of <code>p_value</code> to 0.05, 0.01 and 0.001 respectively to filter the significant DE genes (of course, in applications, you need to correct p-values), apply <code><a href="../reference/ora.html">ora()</a></code> to the three DE gene lists using the GO BP gene sets (think how to obtain the GO gene sets from <strong>org.Hs.eg.db</strong>) and compare the ORA results (the whatever comparison you can think of).</p>
<p>If you EntreZ ID in the gene sets, you need to convert gene symbols for the DE genes to EntreZ IDs. But you can also choose to use gene symbols in the gene sets.</p>
</div>
<div class="section level4">
<h4 id="solution">Solution<a class="anchor" aria-label="anchor" href="#solution"></a>
</h4>
<p>The three DE gene lists:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">de_genes_1</span> <span class="op">=</span> <span class="va">de</span><span class="op">$</span><span class="va">symbol</span><span class="op">[</span><span class="va">de</span><span class="op">$</span><span class="va">p_value</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span></span>
<span><span class="va">de_genes_2</span> <span class="op">=</span> <span class="va">de</span><span class="op">$</span><span class="va">symbol</span><span class="op">[</span><span class="va">de</span><span class="op">$</span><span class="va">p_value</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">]</span></span>
<span><span class="va">de_genes_3</span> <span class="op">=</span> <span class="va">de</span><span class="op">$</span><span class="va">symbol</span><span class="op">[</span><span class="va">de</span><span class="op">$</span><span class="va">p_value</span> <span class="op">&lt;</span> <span class="fl">0.001</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">de_genes_1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 3156</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">de_genes_2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1708</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">de_genes_3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 817</span></span></code></pre>
<p>Get the GO BP gene sets. Note genes in <code>de_genes_*</code> are symbols, thus genes in gene sets should also be symbols.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_GO_gene_sets_from_orgdb.html">get_GO_gene_sets_from_orgdb</a></span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, <span class="st">"BP"</span>, gene_id_type <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'select()' returned 1:many mapping between keys and columns</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gs</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## $`GO:0000002`</span></span>
<span><span class="co">##  [1] "PARP1"    "SLC25A4"  "DNA2"     "TYMP"     "ENDOG"    "LIG3"    </span></span>
<span><span class="co">##  [7] "MEF2A"    "MPV17"    "OPA1"     "POLG"     "RRM1"     "SSBP1"   </span></span>
<span><span class="co">## [13] "TOP3A"    "TP53"     "DNAJA3"   "LONP1"    "AKT3"     "PPARGC1A"</span></span>
<span><span class="co">## [19] "POLG2"    "RRM2B"    "SLC25A36" "TWNK"     "METTL4"   "PIF1"    </span></span>
<span><span class="co">## [25] "SESN2"    "SLC25A33" "MGME1"    "FLCN"     "PRIMPOL"  "STOX1"</span></span></code></pre>
<p>We perform ORA for the three DE gene lists.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tb1</span> <span class="op">=</span> <span class="fu"><a href="../reference/ora.html">ora</a></span><span class="op">(</span><span class="va">de_genes_1</span>, <span class="va">gs</span><span class="op">)</span></span>
<span><span class="va">tb2</span> <span class="op">=</span> <span class="fu"><a href="../reference/ora.html">ora</a></span><span class="op">(</span><span class="va">de_genes_2</span>, <span class="va">gs</span><span class="op">)</span></span>
<span><span class="va">tb3</span> <span class="op">=</span> <span class="fu"><a href="../reference/ora.html">ora</a></span><span class="op">(</span><span class="va">de_genes_3</span>, <span class="va">gs</span><span class="op">)</span></span></code></pre></div>
<p>Compare number of significant GO terms:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">tb1</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1025</span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">tb2</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 871</span></span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">tb3</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 788</span></span></code></pre>
<p>Actually we can see gene set enrichment analysis is a very robust method which is not sensitive to DE cutoffs (at least for this dataset).</p>
<p>Make a Venn (Euler) diagram:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jolars/eulerr" class="external-link">eulerr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://jolars.github.io/eulerr/reference/euler.html" class="external-link">euler</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"DE_cutoff_0.05"</span>  <span class="op">=</span> <span class="va">tb1</span><span class="op">$</span><span class="va">gene_set</span><span class="op">[</span><span class="va">tb1</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span>,</span>
<span>                <span class="st">"DE_cutoff_0.01"</span>  <span class="op">=</span> <span class="va">tb2</span><span class="op">$</span><span class="va">gene_set</span><span class="op">[</span><span class="va">tb2</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span>,</span>
<span>                <span class="st">"DE_cutoff_0.001"</span> <span class="op">=</span> <span class="va">tb3</span><span class="op">$</span><span class="va">gene_set</span><span class="op">[</span><span class="va">tb3</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, </span>
<span>     quantities <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="topic2_02_implement_ora_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Compare log2 fold enrichment from the three ORA results. First take the common GO terms in the three ORA tables.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tb1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tb2</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tb3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Then simply make pairwise scatter plots:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tb1</span><span class="op">[</span><span class="va">cn</span>, <span class="st">"log2fe"</span><span class="op">]</span>, <span class="va">tb2</span><span class="op">[</span><span class="va">cn</span>, <span class="st">"log2fe"</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">16</span>, cex <span class="op">=</span> <span class="fl">0.5</span>, col <span class="op">=</span> <span class="st">"#00000080"</span>,</span>
<span>    xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"compare log2 fold enrichment from ORA"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"DE_cutoff_0.05"</span>, ylab <span class="op">=</span> <span class="st">"DE_cutoff_0.01"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0</span>, b <span class="op">=</span> <span class="fl">1</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tb1</span><span class="op">[</span><span class="va">cn</span>, <span class="st">"log2fe"</span><span class="op">]</span>, <span class="va">tb3</span><span class="op">[</span><span class="va">cn</span>, <span class="st">"log2fe"</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">16</span>, cex <span class="op">=</span> <span class="fl">0.5</span>, col <span class="op">=</span> <span class="st">"#00000080"</span>,</span>
<span>    xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"compare log2 fold enrichment from ORA"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"DE_cutoff_0.05"</span>, ylab <span class="op">=</span> <span class="st">"DE_cutoff_0.001"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0</span>, b <span class="op">=</span> <span class="fl">1</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tb2</span><span class="op">[</span><span class="va">cn</span>, <span class="st">"log2fe"</span><span class="op">]</span>, <span class="va">tb3</span><span class="op">[</span><span class="va">cn</span>, <span class="st">"log2fe"</span><span class="op">]</span>, pch <span class="op">=</span> <span class="fl">16</span>, cex <span class="op">=</span> <span class="fl">0.5</span>, col <span class="op">=</span> <span class="st">"#00000080"</span>,</span>
<span>    xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>, <span class="fl">5</span><span class="op">)</span>, main <span class="op">=</span> <span class="st">"compare log2 fold enrichment from ORA"</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"DE_cutoff_0.01"</span>, ylab <span class="op">=</span> <span class="st">"DE_cutoff_0.001"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0</span>, b <span class="op">=</span> <span class="fl">1</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span></span></code></pre></div>
<p><img src="topic2_02_implement_ora_files/figure-html/unnamed-chunk-10-1.png" width="1152"></p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
